### DISCUSSION TITLE
[dynamic]/[route].js getServerSideProps favicon.ico in params

### DISCUSSION THREAD

**Discussion by ditorodev** [vercel/next.js#13301] [Help] [answered]
Hey again 😅!!

I have this route inside `pages` it is `[customPage]/index.js` on index.js I'm doing the next:

```
export async function getServerSideProps({req, res, params}) {
  console.log(params); // here I'm getting {customPage: 'favicon.ico'}  😕
  const programs_data = await getAllRelatedPages(params.customPage);
  const location = req.headers.host;
  // console.log(`${location}/programas/${programs_data[0].slug}`);
  if(res) {
    res.writeHead(301, {
      Location: `http://${location}/${params.customPage}/${programs_data[0].slug}`
    });
    res.end();
  }

  return {};
};
```
(It is just a simple redirect to an known page)

As you can see in the comment, params.customPage is not what's being passed on the URL(ex: localhost:3000/example).

this is my folder structure: 
```
pages/
-- [customPage]/
---- [slug].js
---- index.js
```

**Comment by kidabu9x** [comment]
I'm facing this problem, too. Has anyone had a way to fix this?

**Reply by ditorodev** [reply]
As I had few routes by the time I just copy and paste the `getServerSideProps` into each new one and change `const programs_data = await getAllRelatedPages(params.customPage);` to my desired customPage. 

But I have an idea(that I won't implement for this project most probably), you can user generators and have an endpoint to generate the templates for this page, in my case I just want to change the parameter to the `getAllRelatedPages` API call, so in theory you just have to create `pages/[customPage(passed thru the endpoint)]/[param].js` `pages/[customPage]/index.js` files and then trigger `nex start && next build` or call your vercel webhook to build the site again. You can use a template engine like [ejs](https://ejs.co/) to generate this routes. It should be pretty straightforward to implement, if you do implement this please let me know 😅

**Comment by remjx** [comment, marked as answer]
I had the same issue and was able to fix it by setting a `/public/favicon.ico` icon (does not exist as part of next.js boilerplate)

**Reply by mohamedsahel** [reply]
Thank you, I really don't know what causes this, but your hack is working 👍

**Reply by Janeqq1** [reply]
Solved my problem too! How in the world did you figure this one out? Impressive!

**Reply by leerob** [reply]
FYI, the default create-next-app template _does_ include this file.

https://github.com/vercel/next.js/tree/canary/packages/create-next-app/templates/default/public

**Reply by bryanprimus** [reply]
Still facing the same issue, I try to add favicon.ico `/public/favicon.ico` but I still get it logged on my middleware params

![image](https://user-images.githubusercontent.com/46083126/160734939-887940fe-9f35-47e0-94f0-5ecea9b92b23.png)

**Reply by josegutierro** [reply]
I'm facing the same issue on my getServerSideProps on folder `pages/[customPage]/[page].js` anyone solved the problem?

**Comment by BenjaminIsMyName** [comment]
I solved it by adding this line to my page (inside `<Head>... </Head>`:

```html
 <link rel="icon" href="/favicon.ico" />
```

Emphasis on the **/** before favicon!

**Reply by Shoreasg** [reply]
This solves my problem for Next13!

**Reply by alexnguyennz** [reply]
Thank you, adding a `/` fixed it for me and removes it from the params object (next@13.0.6).

**Reply by ObadaJasemHolo** [reply]
I was also missing the `/`, that worked, any idea why?

**Reply by JohnMwanyika** [reply]
if it works dont touch it😅

**Reply by rubel-sh** [reply]
thanks bro, it solved my issue. was missing the `/`

**Comment by Shmookoff** [comment]
For app directory users: make sure `favicon.ico` is placed in app directory.
Specifically, my problem was that I had two route groups with multiple root layouts:
```
app/
├─ (cms)/
│  ├─ layout.js
├─ (web)/
│  ├─ layout.js/
```
When converting to this structure, I forgot to place `favicon.ico` in root of `app` directory:
```
--- BAD ---
app/
├─ (cms)/
│  ├─ layout.js
├─ (web)/
│  ├─ layout.js/
│  ├─ favicon.ico
```
```
--- GOOD ---
app/
├─ (cms)/
│  ├─ layout.js
├─ (web)/
│  ├─ layout.js/
├─ favicon.ico
```
https://nextjs.org/docs/app/api-reference/file-conventions/metadata/app-icons#favicon

**Comment by greebl3** [comment]
Updating for page-router users, not [app-router users](https://github.com/vercel/next.js/discussions/13301#discussioncomment-6892785).

The fix mentioned [here](https://github.com/vercel/next.js/discussions/13301#discussioncomment-79557) **partially** works, but has the following caveats:
-  adding `/public/favicon.ico` removes `favicon.ico` from `context.params`
- `context.params.length` **is still greater than 1**

Check your response and/or JS object accordingly

- NextJS: 13.1.6 on Mac OS Monterey 12.5
